### ISSUE TITLE
[VL] Optimize shuffle writer memory usage

### ISSUE THREAD

**Issue by Yohahaha** [oap-project/gluten#3444]
### Description

A simple task with scan + filter + shuffle writer has 45MB of peek memory usage, assume scan and filter belong to computation part, shuffle writer belong to shuffle part, after add detailed log, we find computation part consume 25MB, shuffle part consume 20MB.

I'm not start but I think shuffle part's memory usage could be optimized.

![image](https://github.com/oap-project/gluten/assets/34979747/b9770453-b1c8-41e4-ba96-ea85b813cc76)

set spark.gluten.memory.reservationBlockSize=1MB to minimize reservation.

**Comment by zhouyuan** [comment]
CC: @marin-ma

**Comment by marin-ma** [comment]
Could you try lowering the `spark.gluten.shuffleWriter.bufferSize`? The default value is 4096.

**Comment by Yohahaha** [comment]
> Could you try lowering the `spark.gluten.shuffleWriter.bufferSize`? The default value is 4096.

Yes, shuffle part peek memory decreased from 20MB to 8MB by adjust above value to 1024.

Could you share more about this config? tuning this config will affect performance?

**Comment by marin-ma** [comment]
@Yohahaha We haven't tested performance around adjusting this configuration. For simple SQL queries, you can check the "avg read batch num rows" metrics in `ColumnarShuffle` op from the SQL tab in spark UI, and set this value smaller by referencing the metric value. But for complex SQL queries, this value can vary among different shuffle stages, therefore the configuration is hard to decide, and we suggest using the default value 4096.

By the way, the "Peak memory allocation" here only indicates the virtual memory allocation, not the actually used physical memory.
